include/zephyr/pm/pm.h

```diff
@@ -53,18 +53,136 @@ extern "C" {
  */
 struct pm_notifier {
 	sys_snode_t _node;
+
 	/**
-	 * Application defined function for doing any target specific operations
+	 * Value indicating whether the callback should trigger on entry or
+	 * exit of a power state.
+	 */
+	enum pm_state_direction direction;
+
+	/**
+	 * Application or driver defined function for doing any target specific operations
 	 * for power state entry.
 	 */
-	void (*state_entry)(enum pm_state state);
+	void (*callback)(void *arg1, void *arg2, void *arg3);
```

Comment:
I would change it to something like:
```
void (*callback)(enum pm_state_direction direction, const void *ctx);
```

The reason to just having a single pointer for context is discussed bellow and there is no need to make this struct bigger for places where multiple args are not needed.

Giving the direction is something that I think makes sense, it allows re-using the same callback in case a user needs to do some work before/after a state. So, that means having something like `PM_STATE_ENTRY_AND_EXIT`

Reply:
Agreed on the `const void *ctx` part.

I can see the reasoning behind `PM_STATE_ENTRY_AND_EXIT`, but I think it muddies the purpose of the direction variable a bit, as we would never pass that third value to `pm_state_notify()`.

Reply:
Taking a second to think about it, I think adding `PM_STATE_ENTRY_AND_EXIT` makes sense.